### PR TITLE
Add support for building snaps

### PR DESCRIPTION
**Pull request by igorljubuncic**
Hello,

This PR adds support for building a snap package of tqdm.

Snaps are cross-distro Linux software packages. One snap can be installed on all supported Ubuntu LTS and non LTS releases from 14.04 onward. Additionally, they can installed on Debian, Manjaro, Fedora, OpenSUSE and many others distributions. Making a snap of tqdm enables you to provide automatic updates on your schedule to your users via the snap store. The store is used by an audience of millions, so this could also give you extra exposure to your project.

If you accept this PR, you can use snapcraft locally, a CI system such as travis or circle-ci, or our free build system (build.snapcraft.io) to create snaps and upload to the store (snapcraft.io/store).

To test this PR locally, I used a stock Ubuntu 18.04 system, with the following steps:

snap install snapcraft --classic --beta
git clone https://github.com/igorljubuncic/tqdm.git
cd tqdm
git checkout add-snapcraft
snapcraft

The generates a <tqdm-version>.snap file, something like tqdm_1.0_amd64.snap.

This snap can be installed and tested local with:

snap install tqdm_1.0_amd64.snap --dangerous

The --dangerous flag is necessary because we’re installing an app that hasn’t gone through the snap store review process and isn't signed.

Once installed, the (namespaced) tqdm command can be executed.

If landed, you will need to:

- Register a developer account in the snap store https://snapcraft.io/account.
- Register the tqdm name in the store. 

snapcraft login
snapcraft register

- Upload a built snap to the store.

snapcraft push tqdm_1.0_amd64.snap --release edge

- Test installing on a clean machine

snap install tqdm --edge

The store supports multiple risk levels as “channels” with the 'edge' channel typically used to host the latest build from git master. Stable is where stable releases are pushed. Optionally, beta and candidate channels can also be used if needed.

Once (and if) you are happy, you can push a stable release to the stable channel, update the store page, and promote the application online. We can help there, and we'd be glad to feature your application in our store.

Thanks!

### CODE REVIEW

**Comment by casperdcl** [comment]
awesome. related: #247

File: snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+description: |
+  A fast, extensible progress bar for Python and CLI
+grade: stable
+confinement: strict
+base: core18
+icon: tqdm-logo.png
+parts:
+  tqdm:
+    plugin: python
+    python-version: python2
```
Comment (casperdcl) [inline snapcraft.yaml]: python3 is also supported

File: snapcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+name: tqdm
+version: '1.0'
```
Comment (casperdcl) [inline snapcraft.yaml]: should this be `tqdm`'s version, or is this a `snap` build number?

File: snapcraft.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+summary: A fast, extensible progress bar for Python and CLI
+description: |
+  A fast, extensible progress bar for Python and CLI
```
Comment (casperdcl) [inline snapcraft.yaml]: any way to inject a file (README.rst) into this field?

File: snapcraft.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+grade: stable
+confinement: strict
+base: core18
+icon: tqdm-logo.png
+parts:
+  tqdm:
+    plugin: python
+    python-version: python2
+    source: https://github.com/tqdm/tqdm.git
```
Comment (casperdcl) [inline snapcraft.yaml]: is the `.git` extension required?

File: snapcraft.yaml (hunk 2 above)
Comment (igorljubuncic) [reply snapcraft.yaml]: You can use any version you like - it carries no semantic meaning in snaps. You can make this snap revision or match to your application.

File: snapcraft.yaml (hunk 1 above)
Comment (igorljubuncic) [reply snapcraft.yaml]: Yes, absolutely, you can also build with python3. Unless you use specific python3 stuff, then it's best to build with python2 for max. compatibility.

File: snapcraft.yaml (hunk 3 above)
Comment (igorljubuncic) [reply snapcraft.yaml]: You could perhaps have a script that generates a snapcraft.yaml stub with description already pre-populated, and then add the remaining stanzas. Or maybe pipe readme.rst into sed and then replace a default description with yours.

File: snapcraft.yaml (hunk 4 above)
Comment (igorljubuncic) [reply snapcraft.yaml]: Your source can be anything - online repos, archives, local files. As long as the source is python so the python plugin can build the application correctly.

File: snapcraft.yaml (hunk 1 above)
Comment (casperdcl) [reply snapcraft.yaml]: due to #359 might be best to have python3 only

Comment (igorljubuncic) [reply snapcraft.yaml]: Sure thing, so you simply declare python-version: python3 instead.

File: snapcraft.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+apps:
+  tqdm:
+    command: bin/tqdm
+    plugs:
+      - home
```
Comment (casperdcl) [inline snapcraft.yaml]: is this plug required?

File: snapcraft.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+name: tqdm
+version: '1.0'
+summary: A fast, extensible progress bar for Python and CLI
+description: |
+  A fast, extensible progress bar for Python and CLI
+grade: stable
+confinement: strict
+base: core18
```
[the remark this answers is not in the document]
Comment (igorljubuncic) [reply snapcraft.yaml]: Core18 is the recommended option.

File: snapcraft.yaml (hunk 5 above)
Comment (igorljubuncic) [reply snapcraft.yaml]: Unless the app needs not access anything inside the home dir, then not. You can test and restrict further as you see fit. And to answer another question you raised earlier, there can be multiple binaries, even with different plugs if you choose so.

**Comment by casperdcl** [comment]
Thanks @igorljubuncic - can you check out https://github.com/tqdm/tqdm/tree/snap and see if you can work out why it's not building (error above https://github.com/tqdm/tqdm/pull/647#issuecomment-457863407)?

**Comment by igorljubuncic** [comment]
What snapcraft version do you have installed on your system?
Do you have the final yaml after the dynamic fields are parsed and populated?
I'm interested in how source and source-commit (and a few other fields) look like.

**Comment by casperdcl** [comment]
```yml
name: tqdm
version: '1.0'
version-script: python -m tqdm --version
summary: A fast, extensible progress bar for Python and CLI
description: |
  https://tqdm.github.io
grade: stable
confinement: strict
base: core18
icon: logo.png
parts:
  tqdm:
    plugin: python
    python-version: python3
    source: .
    source-commit: db4452e
apps:
  tqdm:
    command: bin/tqdm
    plugs:
      - home
```

snapcraft version 3.1

**Comment by casperdcl** [comment]
Yes, the [`snap`](https://github.com/tqdm/tqdm/tree/snap) branch

**Comment by casperdcl** [comment]
@igorljubuncic any update?

**Comment by igorljubuncic** [comment]
Sorry, I thought I had replied. I tested this, and was able to build after cleaning (snapcraft clean). Can you check if the same works for you? Might be a partial leftover from one of the previous build attempts that conflicts with this.

Also, I wasn't able to auto-derive the version (I used a static string). But we can figure that one out separately.

**Comment by casperdcl** [comment]
yes, thanks, I added 2486e0b26c841231816d013fd46e4c7e88e2e248 which avoids the issue. Seems to be a snapcraft bug - using `python -m tqdm --version` or `tqdm --version` for the `version-script` key should work...

**Comment by igorljubuncic** [comment]
I'll raise the point with our dev team, and if this is indeed a bug, we will fix it.

**Comment by igorljubuncic** [comment]
Hello,

A couple of suggestions:

Version-script is deprecated, but you might succeed by perhaps using: 
./usr/bin/python3 -m tdqm (or /usr/bin/python3 -m tdqm)

There's this doc on how to use external metadata, so this should help:
https://docs.snapcraft.io/using-external-metadata/4642

Let me know if you have other questions or issues.

**Pull request closed without merging**